scalableminds/webknossos
# Speed up mapping creation/activation

This PR speeds up mapping creation/activation of large mappings substantially (~5-6x faster). Key optimizations are:
- avoid resizing of ArrayBuffers by allocating a buffer padded to the correct size right at the start
- avoid calling Object.keys (and _.size(...) for that matter), as it is really slow for large objects, by keeping track of the keys ourselves

Example: Activating the `astrocyte-full` mapping went from taking 41.1s overall to taking 8.7s overall for me. The time to create the texture went from taking 23.3s to taking 3.8s.

I added a screenshot test for mappings while I was at it :)

### URL of deployed dev instance (used for testing):
- https://speedupmappingcreation.webknossos.xyz

### Steps to test:
- Open a tracing and use the mapping UI to activate a mapping. Everything should work and it should be noticeably faster than on master.
- Use the api to activate a mapping: `webknossos.apiReady().then(async api => api.data.setMapping("segmentation", { <visible_id>: 0}))`

------
